src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaHelper.cs

```diff
@@ -0,0 +1,220 @@
+                }
+
+                if (message is not null && message.Timestamp.Type != 0)
+                {
+                    var consumeTime = span.StartTime.UtcDateTime;
+                    var produceTime = message.Timestamp.UtcDateTime;
+                    var queueTimeMs = Math.Max(0, (consumeTime - produceTime).TotalMilliseconds);
+                    // TODO: should this be a metric or tag?
+                    tags.RecordQueueTimeMs = queueTimeMs.ToString("0");
+                }
+
+                if (message is not null && message.Value is null)
+                {
+                    tags.Tombstone = "true";
+                }
+
+                // Consumer spans should always be measured
+                span.SetTag(Tags.Measured, "1");
+
+                tags.SetAnalyticsSampleRate(KafkaConstants.IntegrationId, tracer.Settings, enabledWithGlobalSetting: false);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+            }
+
+            return scope;
+        }
+
+        internal static void CloseConsumerScope(Tracer tracer)
+        {
+            try
+            {
+                if (!tracer.Settings.IsIntegrationEnabled(KafkaConstants.IntegrationId)
+                    || !tracer.Settings.KafkaCreateConsumerScopeEnabled)
+                {
+                    // integration disabled, skip this trace
+                    return;
+                }
+
+                var activeScope = tracer.ActiveScope;
+                var currentSpan = activeScope?.Span;
+                if (currentSpan?.OperationName != KafkaConstants.ConsumeOperationName)
+                {
+                    // Not currently in a consumer operation
+                    return;
+                }
+
+                // TODO: record end-to-end time?
+                activeScope.Dispose();
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error closing Kafka consumer scope");
+            }
+        }
+
+        /// <summary>
+        /// Try to inject the prop
+        /// </summary>
+        /// <param name="context">The Span context to propagate</param>
+        /// <param name="message">The duck-typed Kafka Message object</param>
+        /// <typeparam name="TTopicPartitionMarker">The TopicPartition type (used  optimisation purposes)</typeparam>
+        /// <typeparam name="TMessage">The type of the duck-type proxy</typeparam>
+        internal static void TryInjectHeaders<TTopicPartitionMarker, TMessage>(SpanContext context, TMessage message)
+            where TMessage : IMessage
+        {
+            // TODO: Allow disabling globally
+            if (!_headersInjectionEnabled)
+            {
+                return;
+            }
+
+            try
+            {
+                var headers = message.Headers ?? CachedMessageHeadersHelper<TTopicPartitionMarker>.CreateAndSetHeaders(message);
```

Comment:
I think there's more coupling here than needed by passing `message`. How about something like this so that `CachedMessageHeadersHelper.CreateHeaders()` is only concerned with creating the headers and not setting them?
```csharp
if(message.Headers == null)
{
    message.Headers = CachedMessageHeadersHelper<TTopicPartitionMarker>.CreateHeaders();
}

var adapter = new KafkaHeadersCollectionAdapter(message.Headers);
```

Reply:
Sure, this method was originally doing a lot more, I'll fix as suggested

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/1444/commits/4a4325cd76b7494fbfd6d1a04afdcab3dd06516c